golang/internal/grpc/grpc.go

```diff
@@ -351,8 +269,118 @@ func (cl *ClientLoop) grpcLoop(connParams *ConnectionParams) {
 	}
 }
 
-func (cl *ClientLoop) handleGrpcTokenError(err error) {
-	if cl.AppConfig.JwtToken.Type == config.Connection {
+func initWithToken(
+	grpcContext context.Context,
+	appConfig *config.CommonConfiguration,
+	workerFuncs *WorkerFunctions,
+	secrets config.SecretStore,
+	token *config.ValidJWT,
+) error {
+	address := token.Issuer
+
+	ctx, cancel := context.WithCancel(grpcContext)
+	loop := ClientLoop{
+		cancel:      cancel,
+		AppConfig:   appConfig,
+		Ctx:         ctx,
+		WorkerFuncs: *workerFuncs,
+		Secrets:     secrets,
+		NodeID:      token.Subject,
+	}
+
+	loop.Ctx = metadata.AppendToOutgoingContext(loop.Ctx, contextMetadataKeyToken, token.StringifiedToken)
+
+	var creds credentials.TransportCredentials
+
+	httpAddr := fmt.Sprintf("https://%s", address)
+	certPool, err := fetchCertificatesFromURL(loop.Ctx, httpAddr)
+	if err != nil {
+		if appConfig.Debug {
+			log.Warn().Err(err).Msg("Secure mode is disabled in demo/dev environment, falling back to plain-text gRPC")
+			creds = insecure.NewCredentials()
+		} else {
+			log.Panic().Err(err).Msg("Could not fetch valid certificate")
+		}
+	} else {
+		creds = credentials.NewClientTLSFromCert(certPool, "")
+	}
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(creds),
+		grpc.WithBlock(),
+		grpc.WithKeepaliveParams(
+			keepalive.ClientParameters{
+				Time:                appConfig.GrpcKeepalive,
+				Timeout:             appConfig.DefaultTimeout,
+				PermitWithoutStream: true,
+			}),
+	}
+
+	log.Info().Str("address", address).Msg("Dialing to address.")
+	conn, err := grpc.Dial(address, opts...)
+	if err != nil {
+		log.Panic().Stack().Err(err).Msg("Failed to dial gRPC")
+	}
+
+	for {
+		state := conn.GetState()
+		if state != connectivity.Ready {
+			log.Debug().Msgf("Waiting for state to change: %d", state)
+			conn.WaitForStateChange(loop.Ctx, state)
+			log.Debug().Msgf("State Changed to: %d", conn.GetState())
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		log.Error().Stack().Err(err).Msg("gRPC connection error")
+	}
+	grpcConn.Conn = conn
+
+	return loop.grpcLoop(token)
+}
+
+func Init(grpcContext context.Context,
+	appConfig *config.CommonConfiguration,
+	secrets config.SecretStore,
+	workerFuncs *WorkerFunctions,
+) {
+	log.Info().Msg("Spinning up gRPC Agent client...")
+	if grpcConn == nil {
+		grpcConn = &Connection{}
+	}
+
+	healthContext, cancel := context.WithCancel(grpcContext)
+	defer cancel()
+
+	err := health.Serve(healthContext)
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to start serving health")
+	}
+
+	err = initWithToken(grpcContext, appConfig, workerFuncs, secrets, appConfig.JwtToken)
+	if err == nil {
+		return
+	}
+
+	if !errors.Is(err, ErrConnectionRefused) {
+		log.Panic().Err(err).Msg("Connection refused")
+	}
+
+	if appConfig.FallbackJwtToken == nil {
+		return
+	}
+
+	log.Warn().Msg("Connection failed, trying fallback token")
+
+	err = initWithToken(grpcContext, appConfig, workerFuncs, secrets, appConfig.FallbackJwtToken)
```

Comment:
You already have appConfig in params, why the `appConfig.FallbackJwtToken` again?

Reply:
Because `initWithToken` needs the actual token that it will use for the connection, so at first it is called with the "normal" token, then the fallback token. So eventhough I have the config, I need to know which token to use.

But if you have a better solution in mind I'm open.